plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,251 @@
+import DataLoader from 'dataloader';
+import pThrottle from 'p-throttle';
+import { ANNOTATION_SLACK_NOTIFICATIONS } from './constants';
+import { SlackNotificationOptions } from './types';
+
+export class SlackNotificationProcessor implements NotificationProcessor {
+  private readonly logger: LoggerService;
+  private readonly directWebhookUrl?: string;
+  private readonly channelWebhookUrl?: string;
+  private readonly catalogClient: CatalogClient;
+  private readonly auth: AuthService;
+  private readonly sendSlackNotifications;
+
+  static fromConfig(
+    config: Config,
+    options: {
+      auth: AuthService;
+      discovery: DiscoveryService;
+      logger: LoggerService;
+      catalogClient: CatalogClient;
+    },
+  ): SlackNotificationProcessor[] {
+    const slackConfig =
+      config.getOptionalConfigArray('notifications.processors.slack') ?? [];
+    return slackConfig.map(c => {
+      const directWebhookUrl = c.getOptionalString('directWebhookUrl');
+      const channelWebhookUrl = c.getOptionalString('channelWebhookUrl');
+      return new SlackNotificationProcessor({
+        directWebhookUrl,
+        channelWebhookUrl,
+        ...options,
+      });
+    });
+  }
+
+  private constructor(options: {
+    auth: AuthService;
+    discovery: DiscoveryService;
+    logger: LoggerService;
+    directWebhookUrl?: string;
+    channelWebhookUrl?: string;
+    catalogClient: CatalogClient;
+  }) {
+    const { auth, catalogClient, directWebhookUrl, channelWebhookUrl, logger } =
+      options;
+    this.logger = logger;
+    this.directWebhookUrl = directWebhookUrl;
+    this.channelWebhookUrl = channelWebhookUrl;
+    this.catalogClient = catalogClient;
+    this.auth = auth;
+
+    // https://api.slack.com/apis/rate-limits
+    const throttle = pThrottle({
+      limit: 10,
+      interval: durationToMilliseconds({ minutes: 1 }),
+    });
+    const throttled = throttle(opts => sendSlackNotification(opts));
+    this.sendSlackNotifications = async (opts: SlackNotificationOptions[]) =>
+      await Promise.all(opts.map(message => throttled(message)));
+  }
+
+  getName(): string {
+    return 'Slack';
+  }
+
+  async processOptions(
+    options: NotificationSendOptions,
+  ): Promise<NotificationSendOptions> {
+    if (!this.channelWebhookUrl || options.recipients.type !== 'entity') {
+      return options;
+    }
+
+    const entityRefs = [options.recipients.entityRef].flat();
+
+    const outbound: SlackNotificationOptions[] = [];
+    await Promise.all(
+      entityRefs.map(async entityRef => {
+        const compoundEntityRef = parseEntityRef(entityRef);
+        // skip users as they are sent direct messages, but allow all other entity kinds
+        // to have a channel id annotation.
+        if (compoundEntityRef.kind === 'user') {
+          return;
+        }
+
+        let destination;
+        try {
+          destination = await this.getChannelFromEntity(entityRef);
```

Comment:
This looks a bit strange, as the destination can be configured in the workflow itself.
That means that it will be always overridden by this logic.

At least it would make sense to keep it as optional, if for some specific entity there should be a different Slack channel then the one which is specified in the workflow itself.

Reply:
The way its intended to be set up is that the destination is not hardcoded in the workflow, but passed in from the notification processor. In the README I suggest setting up the workflow with a variable for the destination.